### ISSUE TITLE
Cache.fetch error in storing HTTP responses

### ISSUE THREAD

**Issue by cwli24** [rails/rails#44982]
### Steps to reproduce

```ruby
      threads[idx] = Thread.new do
        # We'll read from the cache first to see if the data already lives there. 
        response = Rails.cache.fetch( tags[idx], expires_in: 3.hours, race_condition_ttl: 2) do
          HTTP.get(EXTDATA_URL, :params => {:tag => tags[idx]})
        end

        if response.status.success?
          # do some stuff here
        end
      end
    threads.each(&:join) 
```

### Expected behavior
The GET request to the external URL returns a 'content/json' response, which should be stored into the cache under the respective tag name.

### Actual behavior
I get this error:
![image](https://user-images.githubusercontent.com/28714891/165983561-f49e92d5-db0a-4fe3-933a-95444dfaed6d.png)
I should note that I tried it with Concurrent-Ruby's Promise, and it was also this cache.fetch line messing causing issues.
In both cases, to what I can tell, the GET returns with a healthy OK response & associated data, but it seems like the cache just does not like storing it. Can't be a concurrency issue, right, since 3.1 made them thread-safe?

### System configuration
**Rails version**:  7.0.2.4

**Ruby version**: 3.1.2p20

**Comment by SkipKayhil** [comment]
Hey @cwli24, I think this is expected behavior. The cache is attempting to serialize `HTTP::Response` by calling `Marshal.dump` on it, which leads to the error you are seeing. It looks like `HTTP::Response#to_a` may help you as that includes the status and headers and can be dumped.

https://github.com/rails/rails/blob/d82885db11a5072f5359daf930c79c55dac4e284/activesupport/lib/active_support/cache.rb#L979
```
irb(main):001:0> require 'http'
=> true
irb(main):002:0> h = HTTP.get("https://github.com")
=> #<HTTP::Response/1.1 200 OK {"Server"=>"GitHub.com", "Date"=>"Fri, 29 Apr 2022 23:16:46 GMT", "Content-Type"=>"t...
irb(main):003:0> Marshal.dump(h)
(irb):3:in `dump': no _dump_data is defined for class Proc (TypeError)
        from (irb):3:in `<main>'
        from /home/hartley/.cache/asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /home/hartley/.cache/asdf/installs/ruby/3.1.0/bin/irb:25:in `load'
        from /home/hartley/.cache/asdf/installs/ruby/3.1.0/bin/irb:25:in `<main>'
```

**Comment by cwli24** [comment]
@SkipKayhil I'm fine with it in retrospect: that the object you're storing to cache has to be serializable via a `dump` & `load` function. (I guess that's what `:coder` option, alternatively, in config.cache_store is for.)
To follow on this, I don't see much of a warning or conspicuous reference to this fact on either the [guides](https://guides.rubyonrails.org/caching_with_rails.html) or [API ](https://api.rubyonrails.org/classes/ActiveSupport/Cache.html) docs, alerting new Rails users. Should this fact more apparent somewhere?

**Comment by SkipKayhil** [comment]
I do see that [`Cache::Store`](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html) mentions data having to be serializable, but I agree the guide could probably be updated, especially:

> Rails' caching mechanism works great for storing any kind of information.

could maybe become `any kind of serializable information`? Would you like to open a PR (with my suggestion or otherwise)?

**Comment by cwli24** [comment]
Sure, I'll get to this. I'd also mention it under [2.2 ActiveSupport::Cache::Store](https://guides.rubyonrails.org/caching_with_rails.html#activesupport-cache-store) to use the `:coder` option, should they wish to store objects that don't have dump defined, and maybe add that HTTP counter example on [Cache::Store](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html) to highlight it, as I think it's intuitive for programmers to believe they'd be able to.

**Comment by jonathanhefner** [comment]
Fixed by de12c18a991d36607268b323157c0715e938debc.
